client/client.go

```diff
@@ -156,10 +156,123 @@ type tsoRequest struct {
 	dcLocation string
 }
 
+type tsoBatchController struct {
+	maxBatchSize         int
+	maxBatchWaitInterval time.Duration
+	// bestBatchSize is a dynamic size that changed based on the current batch effect.
+	bestBatchSize int
+
+	tsoRequestCh          chan *tsoRequest
+	collectedRequests     []*tsoRequest
+	collectedRequestCount int
+
+	batchStartTime time.Time
+}
+
+func newTSOBatchController(tsoRequestCh chan *tsoRequest, maxBatchSize int, maxBatchWaitInterval time.Duration) *tsoBatchController {
+	return &tsoBatchController{
+		maxBatchSize:          maxBatchSize,
+		maxBatchWaitInterval:  maxBatchWaitInterval,
+		bestBatchSize:         maxBatchSize,
+		tsoRequestCh:          tsoRequestCh,
+		collectedRequests:     make([]*tsoRequest, maxBatchSize+1),
+		collectedRequestCount: 0,
+	}
+}
+
+// fetchAllPendingRequests will start a new round of the batch collecting from the channel.
+// It returns true if everything goes well, otherwise false which should stop the service.
+func (tbc *tsoBatchController) fetchAllPendingRequests(ctx context.Context) bool {
+	var firstTSORequest *tsoRequest
+	select {
+	case <-ctx.Done():
+		return false
+	case firstTSORequest = <-tbc.tsoRequestCh:
+	}
+	// Start to batch when the first TSO request arrives.
+	tbc.batchStartTime = time.Now()
+	tbc.collectedRequestCount = 0
+	tbc.pushRequest(firstTSORequest)
+
+	// This loop is for trying best to collect more requests, so we use `tbc.maxBatchSize` here.
+	for tbc.collectedRequestCount < tbc.maxBatchSize {
+		select {
+		case tsoReq := <-tbc.tsoRequestCh:
+			// No more pending TSO requests, just return.
+			if tsoReq == nil {
+				return true
+			}
+			tbc.pushRequest(tsoReq)
+		default:
+			return true
+		}
+	}
+	return true
+}
+
+func (tbc *tsoBatchController) pushRequest(tsoReq *tsoRequest) {
+	tbc.collectedRequests[tbc.collectedRequestCount] = tsoReq
+	tbc.collectedRequestCount++
+}
+
+// shouldFetchMorePendingRequests returns whether we should fetch more pending TSO requests from the channel.
+func (tbc *tsoBatchController) shouldFetchMorePendingRequests() bool {
+	// TODO: maybe consider the actual load that returns through a TSO response from PD server.
+	return tbc.collectedRequestCount < tbc.maxBatchSize && tbc.maxBatchWaitInterval > 0
+}
+
+// fetchMorePendingRequests fetches more pending TSO requests from the channel.
+func (tbc *tsoBatchController) fetchMorePendingRequests() {
+	// Try to collect `tbc.bestBatchSize` requests, or wait `tbc.maxBatchWaitInterval`.
+	after := time.NewTimer(tbc.maxBatchWaitInterval)
+	for tbc.collectedRequestCount < tbc.bestBatchSize {
+		select {
+		case tsoReq := <-tbc.tsoRequestCh:
+			// No more pending TSO requests, just return.
+			if tsoReq == nil {
+				return
+			}
+			tbc.pushRequest(tsoReq)
+		case <-after.C:
+			return
+		}
+	}
+	after.Stop()
```

Comment:
The timer will not stop if it returns early.

Reply:
+1